apm-agent-plugins/apm-asynchttpclient-plugin/src/main/java/co/elastic/apm/agent/asynchttpclient/AsyncHttpClientInstrumentation.java

```diff
@@ -0,0 +1,153 @@
+
+public class AsyncHttpClientInstrumentation extends ElasticApmInstrumentation {
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return hasSuperType(named("org.asynchttpclient.AsyncHttpClient"));
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("executeRequest")
+            .and(takesArgument(0, named("org.asynchttpclient.Request")))
+            .and(takesArgument(1, named("org.asynchttpclient.AsyncHandler")));
+
+    }
+
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Arrays.asList("http-client", "asynchttpclient");
+    }
+
+    @Override
+    public Class<?> getAdviceClass() {
+        return AsyncHttpClientInstrumentationAdvice.class;
+    }
+
+    public static class AsyncHttpClientInstrumentationAdvice {
+        @Nullable
+        @VisibleForAdvice
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        public static void onBeforeExecute(@Advice.Argument(value = 0, readOnly = true) Request request,
+                                           @Advice.Argument(value = 1, readOnly = false) AsyncHandler asyncHandler,
+                                           @Advice.Local("span") Span span) {
+            if(tracer == null || tracer.getActive() == null) {
+                return;
+            }
+
+            final TraceContextHolder<?> parent = tracer.getActive();
+            span = HttpClientHelper.startHttpClientSpan(parent, request.getMethod(), request.getUri().toFullUrl(), request.getUri().getHost());
+
+            if(span != null) {
+                span.activate();
+                request.getHeaders().add(TraceContext.TRACE_PARENT_HEADER, span.getTraceContext().getOutgoingTraceParentHeader().toString());
+            }
+            if(!request.getHeaders().contains(TraceContext.TRACE_PARENT_HEADER) && parent != null) {
+                request.getHeaders().add(TraceContext.TRACE_PARENT_HEADER, parent.getTraceContext().getOutgoingTraceParentHeader().toString());
+            }
+
+            // TODO: Optimize the new call away
+            asyncHandler = new AsyncHandlerWrapperCreator().wrap(asyncHandler, span, request);
+        }
+
+        @Nullable
+        @VisibleForAdvice
+        @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class)
+        public static void onAfterExecute(@Advice.Local("span") Span span,
+                                          @Advice.Thrown Throwable t) {
+            if(span != null) {
+                span.deactivate();
+            }
+        }
+    }
+
+    public static class AsyncHandlerWrapperCreator {
+        public AsyncHandler<Object> wrap(final AsyncHandler asyncHandler, Span span, Request originalRequest) {
+            return new AsyncHandlerWrapper(asyncHandler, span, originalRequest);
+        }
+
+        private static class AsyncHandlerWrapper implements AsyncHandler {
```

Comment:
The default methods are not overridden which would break the delegation to them. Also, new default methods might be added in future versions of this library. I think it's more feasible to not wrap the `AsyncHandler` but rather to instrument the relevant methods of types which implement `AsyncHanlder`. That also has the advantage that we don't have to use the `HelperClassManager`.

The downside is that we have to traverse the inheritance hierarchy of all classes to determine whether a particular class implements `AsyncHandler`. Therefore, it makes sense to add a classloader matcher like this to skip classloaders which can't load asynchttpclient at all:

```java
@Override
public ElementMatcher.Junction<ClassLoader> getClassLoaderMatcher() {
    return not(isBootstrapClassLoader())
        .and(classLoaderCanLoadClass("org.asynchttpclient.AsyncHandler"));
}
```

Reply:
Hm, but the question then is how to access the span from within the listener 🤔